src/bsoncxx/include/bsoncxx/v_noabi/bsoncxx/stdx/optional.hpp

```diff
@@ -14,93 +14,677 @@
+template <typename From, typename To>
+struct enable_opt_value_conversion   //
+    : bsoncxx::detail::conjunction<  //
+          std::is_constructible<To, From&&>,
+          bsoncxx::detail::negation<bsoncxx::detail::is_alike<From, in_place_t>>,
+          bsoncxx::detail::disjunction<
+              bsoncxx::detail::negation<bsoncxx::detail::is_alike<To, bool>>,  //
+              detail::not_an_optional<bsoncxx::detail::remove_cvref_t<From>>>> {};
 
-#elif defined(BSONCXX_POLY_USE_STD_EXPERIMENTAL)
+}  // namespace detail
 
-#include <experimental/optional>
+template <typename T>
+class optional : bsoncxx::detail::equality_operators,
+                 bsoncxx::detail::ordering_operators,
+                 public detail::optional_base_class<T>::type {
+   public:
+    /// The type of value held within this optional
+    using value_type = T;
+    /// An lvalue-reference-to-mutable T
+    using reference = bsoncxx::detail::add_lvalue_reference_t<T>;
+    /// An lvalue-reference-to-const T
+    using const_reference =
+        bsoncxx::detail::add_lvalue_reference_t<bsoncxx::detail::add_const_t<T>>;
+    /// An rvalue-reference-to-mutable T
+    using rvalue_reference = bsoncxx::detail::add_rvalue_reference_t<T>;
+    /// An rvalue-reference-to-const T
+    using const_rvalue_reference =
+        bsoncxx::detail::add_rvalue_reference_t<bsoncxx::detail::add_const_t<T>>;
+    /// A pointer-to-mutable T
+    using pointer = bsoncxx::detail::add_pointer_t<T>;
+    /// A pointer-to-const T
+    using const_pointer = bsoncxx::detail::add_pointer_t<const T>;
 
-namespace bsoncxx {
-namespace v_noabi {
-namespace stdx {
+    // Constructors [1]
+    optional() = default;
+    constexpr optional(nullopt_t) noexcept {}
 
-using ::std::experimental::make_optional;
-using ::std::experimental::nullopt;
-using ::std::experimental::nullopt_t;
-using ::std::experimental::optional;
+    // Ctor [2] and [3] are provided by base classes
+    optional(const optional&) = default;
+    optional(optional&&) = default;
+    // Same with assignments
+    optional& operator=(const optional&) = default;
+    optional& operator=(optional&&) = default;
+    ~optional() = default;
 
-}  // namespace stdx
-}  // namespace v_noabi
-}  // namespace bsoncxx
+    // In-place constructor
+    template <typename... Args>
+    bsoncxx_cxx14_constexpr explicit optional(in_place_t, Args&&... args) noexcept(
+        noexcept(T(BSONCXX_FWD(args)...))) {
+        this->emplace(BSONCXX_FWD(args)...);
+    }
 
-#elif defined(BSONCXX_POLY_USE_STD)
+    // Explicit converting constructor. Only available if implicit conversion is
+    // not possible.
+    template <
+        typename U = T,
+        bsoncxx::detail::requires_t<int,
+                                    detail::enable_opt_value_conversion<U&&, T>,
+                                    bsoncxx::detail::negation<std::is_convertible<U&&, T>>> = 0>
+    bsoncxx_cxx14_constexpr explicit optional(U&& arg) noexcept(
```

Comment:
Does this constructor correspond to [5] on [cppreference](https://en.cppreference.com/w/cpp/utility/optional/optional)?

If "yes", should `explicit` be removed to match? Is the non-constexpr form of [5] needed?

```c++
template < class U >
optional( optional<U>&& other )
```